### PR TITLE
Fix stack trace

### PR DESCRIPTION
**Pull request by ampresent**

### CODE REVIEW

**Comment by Dengerwei** [comment]
This PR seems a little complicated.

For 4.19, __save_stack_trace functions are not optimized in both x86 and arm64, so it can be used directly instead of using sidecar. 

For 5.10, arch_stack_walks are not optimized too in both x86 and arm64,  and as for stack_trace_consume_entry, it can be rewritten again in plugsched.

File: src/stack_check.h [hunk 1]
```diff
@@ -40,14 +44,19 @@ static void stack_check_init(void)
 	#undef EXPORT_PLUGSCHED
 	#undef EXPORT_CALLBACK
 
+	kallsyms_lookup_size_offset((unsigned long)__orig___schedule,
+				&orig_schedule_size, NULL);
+	kallsyms_lookup_size_offset((unsigned long)__mod___schedule,
+				&mod_schedule_size, NULL);
+
```
Comment (Dengerwei) [inline src/stack_check.h:51]: Here, orig_schedule_size and mod_schedule_size variables can be reused the orig___schedule_size and mod___schedule_size defined head_jump.h which need not get the value again.

File: src/stack_check.h [hunk 2]
```diff
@@ -57,41 +66,105 @@ static int stack_check_fn(unsigned long *entries, unsigned int nr_entries, bool
...
+{
+	if (!install && addr >= (unsigned long)__mod___schedule &&
+	    addr < (unsigned long)__mod___schedule + mod_schedule_size)
+		return true;
+	if (install && addr >= (unsigned long)__orig___schedule &&
+	    addr < (unsigned long)__orig___schedule + orig_schedule_size)
+		return true;
+	return false;
+}
+
```
Comment (Dengerwei) [inline src/stack_check.h:149]: Maybe there have another way to skip __schedule. For example, the variable of address of __schedule can be set 0 or ~0, then after address-sorting, the variable is placed at far left or far right, and the bsearch can help to skip this function.

File: src/stack_check.h [hunk 3]
```diff
@@ -57,41 +66,105 @@ static int stack_check_fn(unsigned long *entries, unsigned int nr_entries, bool
...
+	unsigned long address, entries[MAX_STACK_ENTRIES];
+	unsigned int nr_entries;
 
-	return stack_check_fn(entries, nr_entries, install);
+	nr_entries = get_stack_trace(task, entries, MAX_STACK_ENTRIES);
+
+	for (i = 0; i < nr_entries; i++) {
+		address = entries[i];
+		if (is___schedule(address, install))
+			continue;
```
Comment (Dengerwei) [inline src/stack_check.h:161]: See the last comment.

File: src/stack_check.h (hunk 2 above)
Comment (Dengerwei) [reply src/stack_check.h:149]: The size can be set zero, too.

Comment (ampresent) [reply src/stack_check.h:149]: Setting vm_func_size[NR___schedule] and mod_func_size[NR___schedule] to 0 is more like a general and efficient way to skip checking some functions. Bravo.

**Comment by Dengerwei** [comment]
The definitions of __save_stack_trace in x86 and arm64 are different and the usage of it in arm64 is wrong.

File: src/stack_check.h [hunk 4]
```diff
@@ -71,26 +73,91 @@ static int stack_check_fn(unsigned long *entries, unsigned int nr_entries, bool
...
+
+        trace.skip = 0;
+        trace.nr_entries = 0;
+        trace.max_entries = MAX_STACK_ENTRIES;
+        trace.entries = store;
+
+	if (!try_get_task_stack(tsk))
+		return 0;
+
+	__save_stack_trace(&trace, tsk, NULL, false);
```
Comment (Dengerwei) [inline src/stack_check.h:149]: Using save_stack_trace?

**Comment by Dengerwei** [comment]
LGTM

**Pull request merged**